crates/driver/src/tests/cases/jit_orders.rs

```diff
@@ -0,0 +1,183 @@
+use crate::{
+    domain::{
+        competition::{order, order::Side},
+        eth,
+    },
+    tests::{
+        self,
+        cases::{is_approximately_equal, EtherExt},
+        setup::{
+            self,
+            ab_adjusted_pool,
+            ab_liquidity_quote,
+            ab_order,
+            ab_solution,
+            test_solver,
+            Test,
+        },
+    },
+};
+
+struct Amounts {
+    sell: eth::U256,
+    buy: eth::U256,
+}
+
+struct Execution {
+    // The executed net-amounts (including network fee) reported by the solver
+    solver: Amounts,
+    // The executed net-amounts (including network and protocol) reported by the driver
+    driver: Amounts,
+}
+
+struct Order {
+    sell_amount: eth::U256,
+    buy_amount: eth::U256,
+    side: order::Side,
+}
+
+struct JitOrder {
+    order: Order,
+}
+
+struct Solution {
+    jit_order: JitOrder,
+    expected_score: eth::U256,
+}
+
+struct TestCase {
+    order: Order,
+    execution: Execution,
+    is_surplus_capturing_jit_order: bool,
+    solution: Solution,
+}
+
+#[cfg(test)]
+async fn protocol_fee_test_case(test_case: TestCase) {
+    let test_name = format!("JIT Order: {:?}", test_case.order.side);
+    // Adjust liquidity pools so that the order is executable at the amounts
+    // expected from the solver.
+    let quote = ab_liquidity_quote()
+        .sell_amount(test_case.execution.solver.sell)
+        .buy_amount(test_case.execution.solver.buy);
+    let pool = ab_adjusted_pool(quote);
+    let solver_fee = test_case.execution.driver.sell / 100;
+
+    let jit_order = setup::JitOrder {
+        order: ab_order()
+            .kind(order::Kind::Limit)
+            .sell_amount(test_case.solution.jit_order.order.sell_amount)
+            .buy_amount(test_case.solution.jit_order.order.buy_amount)
+            .solver_fee(Some(solver_fee))
+            .side(test_case.solution.jit_order.order.side)
+            .no_surplus(),
+    };
+
+    let order = ab_order()
+        .kind(order::Kind::Limit)
+        .sell_amount(test_case.order.sell_amount)
+        .buy_amount(test_case.order.buy_amount)
+        .solver_fee(Some(solver_fee))
+        .side(test_case.order.side)
+        .partial(0.into())
+        .no_surplus();
+
+    let solver = test_solver();
+    let test: Test = tests::setup()
+        .name(test_name)
+        .pool(pool)
+        .jit_order(jit_order.clone())
+        .order(order.clone())
+        .solution(ab_solution())
+        .surplus_capturing_jit_order_owners(
+            test_case
+                .is_surplus_capturing_jit_order
+                .then(|| vec![solver.address()])
```

Comment:
why do we add solver address here?

Reply:
@sunce86 because in the test the solver is the one doing the JIT order